### DISCUSSION TITLE
How to solve this error java.lang.OutOfMemoryError: Java heap space

### DISCUSSION THREAD

**Discussion by ODOU** [iiasa/message_ix#694] [MESSAGE modeling] [answered]
Hi all, 
I am getting this error **java.lang.OutOfMemoryError: Java heap space** while running **scenario.commit(comment)** command. I suspect the model I am building is a bit large (Regional with 15 nodes [countries]). The exported scenario file in csv is about 50MB. Is there anyway to expand the allocated memory on one hand and on the other hand how to make the model solve faster while launching **scenario.solve()** ?

![image](https://user-images.githubusercontent.com/61684930/222285124-edfe975c-da33-45d9-ae57-2c4fa2404793.png)

![image](https://user-images.githubusercontent.com/61684930/222285293-0914018b-34f4-48f6-a874-538e7b4b0e4d.png)

I will highly appreciate the support of the community. Regards,

**Comment by glatterf42** [comment]
Hi Odou,
an error like yours usually means on of two things happened: either you didn't assign enough heap space to or there was too much memory consumed on the program side. This might be due to bad programming (i.e. a memory leak) or due to incorrect use (e.g. trying to store lots and lots of objects in heap memory without ever releasing them). From what I've read, 90% of the time the issue is one the code side. However, since message_ix and ixmp are in extensive use, I'm not convinced that there is a memory leak issue in a standard function like committing to the database. So you might want to try to increase your heap space, but also: "In [509]" from your screenshot sounds like a lot of input lines. If you do create and commit a lot of scenarios in the notebook(or code file) you get the error from, it might be worth investigating if you could move some of these scenarios to a new notebook. This would allow you to close the connection to the database in one notebook and open a new one in the new notebook, which might release stress from your heap space. 
A lot of this answer is paraphrased from [stackoverflow](https://stackoverflow.com/a/1596040), so I'd recommend reading that answer, too. They further recommend profiling your program and you might be able to do so following the links provided there. However, you could try my suggestions first. Instructions on how to increase heap space can also be found in the answer I linked here. 

Please let us know if this solves your issue :)

**Comment by khaeru** [comment, marked as answer]
Underlying the message_ix package is another one called ixmp that handles data storage. By default, when using current versions of message_ix and ixmp, you are using something called the ixmp JDBCBackend, which is built in Java.

The [`ixmp.JDBCBackend` documentation](https://docs.messageix.org/projects/ixmp/en/stable/api-backend.html#ixmp.backend.jdbc.JDBCBackend) gives some examples of how to pass [arguments for starting the Java Virtual Machine (JVM)](https://docs.messageix.org/projects/ixmp/en/stable/api-backend.html#ixmp.backend.jdbc.start_jvm); the one particular example is about increasing the Java heap space (`-Xmg4G` means to allocate 4 GB of memory for the JVM; this is fully explained in the further linked Java documentation).

Practically, you can cause this to always happen by modifying your [ixmp configuration file](https://docs.messageix.org/projects/ixmp/en/stable/api.html#configuration) (usually somewhere like $HOME/.local/share/ixmp/config.json). If you see:
```json
{
  "platform": {
    "default": "local",
    "local": {
      "class": "jdbc",
      "driver": "hsqldb",
      "path": "/home/username/.local/share/ixmp/localdb/default"
    }
  }
}
```
…then edit to add the jvmargs, for example like:
```json
{
  "platform": {
    "default": "local",
    "local": {
      "class": "jdbc",
      "driver": "hsqldb",
      "jvmargs": "-Xmx10G",
      "path": "/home/username/.local/share/ixmp/localdb/default"
    }
  }
}
```

This is equivalent to passing jvmargs="-Xmx10G" every time you call [`ixmp.Platform(name="local", …)`](https://docs.messageix.org/projects/ixmp/en/stable/api.html#ixmp.Platform).

Of course, the correct choice of heap size depends on your model needs, system available memory, and other factors; there is no one specific size.

**Reply by ODOU** [reply]
Thank @khaeru and @glatterf42  for the support. I have located the config.json file and make the changes as required and it works perfectly.  I am a bit stuck at the solving level  (I passed the demo license limit). What type of academic license type will work best given the size of the model I am working on. See MESSAGE_run.log below :

    +++ Solve the perfect-foresight version of MESSAGEix +++
--- Generating LP model MESSAGE_LP
--- MESSAGE_run.gms(4450) 255 Mb
---   26,974 rows  84,354 columns  245,156 non-zeroes
*** The model exceeds the demo license limits for linear models of more than 2000 rows or columns
*** Status: Terminated due to a licensing error
***         License file: C:\Users\Thierry Odou\Documents\GAMS\gamslice.txt
***         Inspect listing file for more information.
--- Job MESSAGE_run.gms Stop 03/02/23 23:07:58 elapsed 0:00:02.750

Best,

**Reply by khaeru** [reply]
Great, glad to hear that setting worked.

On the question of GAMS licensing, I am not sure. A first step would be to check with colleagues at any institution(s) where you are affiliated and see if they already have licensed GAMS.

Work within the IIASA ECE program almost exclusively uses the CPLEX solver. [Per the price list](https://www.gams.com/sales/pricing_academic/), a perpetual license to a single named user for this solver is currently USD 1400 on top of USD 700 for the base GAMS license, but there are discounts for subscription, multi-user, and others.

We recognize this is a substantial barrier to building large models in `message_ix`; we have talked about enabling e.g. the free and open source [HiGHS solver](https://gams.com/latest/docs/S_HIGHS.html) as an alternative, which would reduce minimum cost to that of the base license (USD 700). However, no work on this has been done.

**Reply by ODOU** [reply]
Hi @khaeru, unfortunately my university does not have provision for CPLEX. While in the meantime, I am looking for  a solution, if I was to use HiGHS what would be requirement and how long it can take to implement it ?

**Reply by khaeru** [reply]
> to use HiGHS what would be requirement and how long it can take to implement it ?

To be clear when I say “we have *talked* about” I mean “we have not actually *tried* (or succeeded).” In trying, we might figure out what kind of adaptations would be needed (if any), and then one could guess how long those might take.

As for *trying*, it's a one-line change: https://github.com/iiasa/message_ix/blob/139c832a9fc195de97e2b4b911461fb49b6fa455/message_ix/model/MESSAGE/auxiliary_settings.gms#L23

If you do try, please share your experience in a fresh discussion thread, rather than this one about Java heap space.

P.S. I am not sure if the limit on the LP problem size for the GAMS free/trial license is for CPLEX in particular, or for any solver. In the latter case, this switch would not solve the problem.

**Reply by brendandanaher** [reply]
Thank you @khaeru. Modifying the config.json file worked very well for me at first. However, when running my code, over time it eventually gets slower and slower every time I re-initialize the scenario as seen in the following code block. 

```jupyter-notebook
import message_ix
import ixmp as ix
from ixmp import Scenario
from message_ix import make_df
#from functions.format_demand import format_demand
import pandas as pd
from pathlib import Path
from message_ix.reporting import Reporter
from update_units import unit_correction
import numpy as np


# Stand-alone model
modelName = 'MESSAGEix_CA_Hydrogen'      # A MESSAGE model name
scenarioName = 'Carbon Price'       # A MESSAGE model scenario name
ssp = 'SSP1'

mp = ix.Platform(name="p1")

# Creating a new MESSAGE stand-alone model
msg = message_ix.Scenario(mp, modelName, scenarioName,
                            version='new', annotation='')
``` 
I've gotten to the point where it takes 20 minutes to run this simple segment of code (this took 30 seconds originally) and I have "jvmargs" set to -Xmx12G in my config.json file. If anyone has a solution to this issue that would be much appreciated. I have attempted clearing memory and temporary files etc with no success. Not sure what else I could try.

[config.json](https://github.com/iiasa/message_ix/files/15030067/config.json)

**Reply by glatterf42** [reply]
Moved this to #825 since it's a new topic :)

**Reply by meng25meng** [reply]
I have the same problem, about the solver, have you solved this problem

**Reply by SongminYu** [reply]
Hi everyone who encounters this error message, I met the same problem and now it is "solved". Maybe this could also help so I leave my steps here.
 - First, I don't know the reason but I don't find the config file. I am using ixmp3.8 with Macbook. But I managed to edit it in PyCharm (Help > Edit Custom VM Options....). However, after doing this, I still see the problem.
 - From the other discussion (#825), I saw recommendations by @khaeru to use several platforms/databases. I tried but it seems I can only use "local" or "default", which I think are the same database?
 - In the end, I found the database folder `user/.local/share/ixmp/localdb` and removed everything in it (3GB+). This "solved" the problem.
